### PR TITLE
[mltest] LearnLog: new ML test case

### PR DESCRIPTION
**Pull request by ZchiPitt**
hi. a few changes i made since my last try:
1). Systematically generated training/testing data, instead of randomly. (Trains faster with simpler network)
2). Use 2 fc only. One with 4 output depth, one with 1 output depth. Used an tanh activation function in between. (I have tried with sigmoid, but worse than tanh)
3). Training data is still sampled from [0.75,1.5], with testing data sampled from [1.1, 1.2]. 
4). The execution time of this MLTest is much shorter than last try. 
- Interpreter/MLTest (86 ms total)
- JIT/MLTest (386 ms total)
- OpenCL/MLTest (948 ms total)
Short than existing test like "learnXor"

### CODE REVIEW

File: tests/unittests/MLTest.cpp [hunk 1]
```diff
@@ -199,6 +199,79 @@ TEST_P(MLTest, learnXor) {
...
+
+  EE_.compile(CompilationMode::Infer, F);
+
+  // Set the testing data.
+  Tensor testSet(ElemKind::FloatTy, {batchSize, 1});
+
+  auto TES = testSet.getHandle<>();
+
+  const float LO_T = 1.1; // Lower bound of testing data.
+  const float HI_T = 1.2; // Upper bound of testing data.
```
Comment (bertmaher) [inline tests/unittests/MLTest.cpp]: Why take test data from a smaller range than train data?

File: tests/unittests/MLTest.cpp [hunk 2]
```diff
@@ -199,6 +199,79 @@ TEST_P(MLTest, learnXor) {
...
+
+  for (size_t i = 0; i < batchSize; i++) {
+
```
Comment (bertmaher) [inline tests/unittests/MLTest.cpp]: Nit, extraneous blank line.

File: tests/unittests/MLTest.cpp [hunk 3]
```diff
@@ -199,6 +199,79 @@ TEST_P(MLTest, learnXor) {
...
+    // Generate a floating number in the range of [LO_T,HI_T).
+    float a = LO_T + i * (HI_T - LO_T) / numInputs;
```
Comment (bertmaher) [inline tests/unittests/MLTest.cpp]: Does this test work with random test data?

File: tests/unittests/MLTest.cpp (hunk 1 above)
Comment (ZchiPitt) [reply tests/unittests/MLTest.cpp]: i actually want to have different range for testing data, and thus i have two more variables for specifying boundaries.  The test case can also work with the same range of the training data.

File: tests/unittests/MLTest.cpp (hunk 3 above)
Comment (ZchiPitt) [reply tests/unittests/MLTest.cpp]: I tried with random test data now. It can work.

**Comment by ZchiPitt** [comment]
This network works pretty well within this input domain, and it's relatively simpler than my previous tries. Does not take long time to train, and produces good estimate within the range [0.75, 1.5].

File: tests/unittests/MLTest.cpp [hunk 4]
```diff
@@ -199,6 +199,78 @@ TEST_P(MLTest, learnXor) {
     EXPECT_NEAR(resH.at({i, 0}), (a ^ b), 0.1);
   }
 }
+/// Learn the logarithmic function.
```
Comment (bertmaher) [inline tests/unittests/MLTest.cpp]: Blank line between functions.

File: tests/unittests/MLTest.cpp [hunk 5]
```diff
@@ -199,6 +199,78 @@ TEST_P(MLTest, learnXor) {
...
+  auto *result = F->createSave("ret", O);
+
+  // Set the training data.
+  Tensor trainingSet(ElemKind::FloatTy, {numInputs, 1});
+  Tensor trainingLabels(ElemKind::FloatTy, {numInputs, 1});
+
+  auto TS = trainingSet.getHandle<>();
+  auto TL = trainingLabels.getHandle<>();
+
+  // Set the training date as floating number from range [0.75, 1.5).
```
Comment (bertmaher) [inline tests/unittests/MLTest.cpp]: Typo: date->data.  Also remove the extra blank line below.

File: tests/unittests/MLTest.cpp [hunk 6]
```diff
@@ -199,6 +199,78 @@ TEST_P(MLTest, learnXor) {
...
+    float a = LO + i * (HI - LO) / numInputs;
+    TS.at({i, 0}) = a;
+    TL.at({i, 0}) = std::log(a);
+  }
+
+  Function *TF = glow::differentiate(F, EE_.getConfig());
+  EE_.compile(CompilationMode::Train, TF);
+
+  // Train the network:
+  EE_.runBatch(1000, {A, Ex}, {&trainingSet, &trainingLabels});
```
Comment (nadavrot) [inline tests/unittests/MLTest.cpp]: Do we really need a 1000 iterations? Can we reduce this to a smaller number?

File: tests/unittests/MLTest.cpp (hunk 4 above)
Comment (ZchiPitt) [reply tests/unittests/MLTest.cpp]: done

File: tests/unittests/MLTest.cpp (hunk 5 above)
Comment (ZchiPitt) [reply tests/unittests/MLTest.cpp]: done

File: tests/unittests/MLTest.cpp (hunk 6 above)
Comment (ZchiPitt) [reply tests/unittests/MLTest.cpp]: the smallest number i tried that can pass the test, is 800. It seems difficult to reduce it to even smaller.  

I might have to modify other parameters so that each batch may learn faster, if we really want to reduce batch number. However the learning rate is relatively high now(0.07), there does not seem to have much space for radical changes. (Please correct me if im wrong)

Also the run time of 800 batches and 1000 batches do not have much differences (~100ms in total). Maybe we just stick to this batch number?

**Comment by ZchiPitt** [comment]
Done with all the changes. Do I land this diff? (Or let me know if you have other concerns)

**Comment by nadavrot** [comment]
@ZchiPitt Congrats on landing this test. Now, let's quantize it.
